device/mellanox/x86_64-mlnx_msn2700-r0/plugins/sfplpmset.py

```diff
@@ -69,101 +66,91 @@ def get_log_ports(handle, sfp_module):
-    meta.swid = SWITCH_ID
-    return meta
-
-def set_sfp_admin_status(sfp_module, admin_status):
-    # Get PMAOS
-    pmaos = ku_pmaos_reg()
-    pmaos.module = sfp_module
-    meta = init_sx_meta_data()
-    meta.access_cmd = SXD_ACCESS_CMD_GET
-    rc = sxd_access_reg_pmaos(pmaos, meta, REGISTER_NUM, None, None)
-    assert rc == SXD_STATUS_SUCCESS, "sxd_access_reg_pmaos failed, rc = %d" % rc
-
-    # Set admin status to PMAOS
-    pmaos.ase = PMAOS_ASE
-    pmaos.ee = PMAOS_EE
-    pmaos.e = PMAOS_E
-    pmaos.rst = PMAOS_RST
-    if admin_status == SX_PORT_ADMIN_STATUS_DOWN:
-        pmaos.admin_status = PMAOS_DISABLE
+def mgmt_phy_mod_pwr_attr_set(handle, module_id, power_attr_type, admin_pwr_mode):
+    sx_mgmt_phy_mod_pwr_attr = sx_mgmt_phy_mod_pwr_attr_t()
+    sx_mgmt_phy_mod_pwr_mode_attr = sx_mgmt_phy_mod_pwr_mode_attr_t()
+    sx_mgmt_phy_mod_pwr_attr.power_attr_type = power_attr_type
+    sx_mgmt_phy_mod_pwr_mode_attr.admin_pwr_mode_e = admin_pwr_mode
+    sx_mgmt_phy_mod_pwr_attr.pwr_mode_attr = sx_mgmt_phy_mod_pwr_mode_attr
+    sx_mgmt_phy_mod_pwr_attr_p = new_sx_mgmt_phy_mod_pwr_attr_t_p()
+    sx_mgmt_phy_mod_pwr_attr_t_p_assign(sx_mgmt_phy_mod_pwr_attr_p, sx_mgmt_phy_mod_pwr_attr)
+    try:
+        rc = sx_mgmt_phy_mod_pwr_attr_set(handle, SX_ACCESS_CMD_SET, module_id, sx_mgmt_phy_mod_pwr_attr_p)
+        assert SX_STATUS_SUCCESS == rc, "sx_mgmt_phy_mod_pwr_attr_set failed"
+    finally:
+        delete_sx_mgmt_phy_mod_pwr_attr_t_p(sx_mgmt_phy_mod_pwr_attr_p)
+
+def mgmt_phy_mod_pwr_attr_get(handle, module_id, power_attr_type):
+    sx_mgmt_phy_mod_pwr_attr_p = new_sx_mgmt_phy_mod_pwr_attr_t_p()
+    sx_mgmt_phy_mod_pwr_attr = sx_mgmt_phy_mod_pwr_attr_t()
+    sx_mgmt_phy_mod_pwr_attr.power_attr_type = power_attr_type
+    sx_mgmt_phy_mod_pwr_attr_t_p_assign(sx_mgmt_phy_mod_pwr_attr_p, sx_mgmt_phy_mod_pwr_attr)
+    try:
+        rc = sx_mgmt_phy_mod_pwr_attr_get(handle, module_id, sx_mgmt_phy_mod_pwr_attr_p)
+        assert SX_STATUS_SUCCESS == rc, "sx_mgmt_phy_mod_pwr_attr_get failed"
+        sx_mgmt_phy_mod_pwr_attr = sx_mgmt_phy_mod_pwr_attr_t_p_value(sx_mgmt_phy_mod_pwr_attr_p)
+        pwr_mode_attr = sx_mgmt_phy_mod_pwr_attr.pwr_mode_attr
+        return pwr_mode_attr.admin_pwr_mode_e, pwr_mode_attr.oper_pwr_mode_e
+    finally:
+        delete_sx_mgmt_phy_mod_pwr_attr_t_p(sx_mgmt_phy_mod_pwr_attr_p)
+
+
+def pwr_attr_set(handle, module_id, ports, attr_type, power_mode):
+    # bring the port down
+    for port in ports:
+        set_port_admin_status_by_log_port(handle, port, SX_PORT_ADMIN_STATUS_DOWN)
+    # set the desired power mode
+    mgmt_phy_mod_pwr_attr_set(handle, module_id, attr_type, power_mode)
+    # get the current port power mode and make sure that it is as per the mode set.
+    admin_pwr_mode, oper_pwr_mode = mgmt_phy_mod_pwr_attr_get(handle, module_id, attr_type)
+    assert power_mode == admin_pwr_mode, "power mode mismatch"
+    # bring the port up
+    for port in ports:
```

Comment:
Do we need a try...finally here in case there is any exception?

Reply:
fixed